cdap-formats/src/main/java/co/cask/cdap/format/io/JsonStructuredRecordDatumReader.java

```diff
@@ -0,0 +1,161 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.format.io;
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.common.io.Decoder;
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link StructuredRecordDatumReader} that decodes from Json.
+ */
+public class JsonStructuredRecordDatumReader extends StructuredRecordDatumReader {
+
+  private static final Map<Schema.Type, JsonToken> SCHEMA_TO_JSON_TYPE = new EnumMap<>(
+    ImmutableMap.<Schema.Type, JsonToken>builder()
+      .put(Schema.Type.NULL, JsonToken.NULL)
+      .put(Schema.Type.BOOLEAN, JsonToken.BOOLEAN)
+      .put(Schema.Type.INT, JsonToken.NUMBER)
+      .put(Schema.Type.LONG, JsonToken.NUMBER)
+      .put(Schema.Type.FLOAT, JsonToken.NUMBER)
+      .put(Schema.Type.DOUBLE, JsonToken.NUMBER)
+      .put(Schema.Type.STRING, JsonToken.STRING)
+      .put(Schema.Type.BYTES, JsonToken.BEGIN_ARRAY)
+      .put(Schema.Type.ARRAY, JsonToken.BEGIN_ARRAY)
+      .put(Schema.Type.MAP, JsonToken.BEGIN_OBJECT)
+      .put(Schema.Type.RECORD, JsonToken.BEGIN_OBJECT)
+      .build()
+  );
+
+  @Override
+  public StructuredRecord read(Decoder decoder, Schema sourceSchema) throws IOException {
+    if (!(decoder instanceof JsonDecoder)) {
+      throw new IOException("The JsonStructuredRecordDatumReader can only decode using a JsonDecoder");
+    }
+
+    return super.read(decoder, sourceSchema);
+  }
+
+  @Override
+  protected String decodeEnum(Decoder decoder, Schema schema) throws IOException {
+    return getJsonReader(decoder).nextString();
+  }
+
+  @Override
+  protected Collection<?> decodeArray(Decoder decoder, Schema elementSchema) throws IOException {
+    List<Object> array = new ArrayList<>();
+
+    JsonReader jsonReader = getJsonReader(decoder);
+    jsonReader.beginArray();
+    while (jsonReader.peek() != JsonToken.END_ARRAY) {
+      array.add(decode(decoder, elementSchema));
+    }
+    jsonReader.endArray();
+
+    return array;
+  }
+
+  @Override
+  protected Map<?, ?> decodeMap(Decoder decoder, Schema keySchema, Schema valueSchema) throws IOException {
+    if (!keySchema.isCompatible(Schema.of(Schema.Type.STRING))) {
+      throw new IOException("Complex key type not supported: " + keySchema);
```

Comment:
may not be immediately obvious to users what 'key' refers to, a little better to say 'Complex key type in maps are not supported: '